### DISCUSSION TITLE
Need for pyvista.DataSet.triangulate with concave polygons

### DISCUSSION THREAD

**Discussion by T4mmi** [pyvista/pyvista#2398] [Q&A] [answered]
The `triangulate()` method is really helpful,
but there is a need for a safe triangulation of concave (non-convex) polygons
![image](https://user-images.githubusercontent.com/6515799/160804816-67a84384-f83e-4807-a03d-e6d0ed27c275.png)
![image](https://user-images.githubusercontent.com/6515799/160804994-f5130abd-f8ab-4298-971a-5b8ce3bc24a7.png)

**Comment by akaszynski** [comment, marked as answer]
I'd use [delaunay_2d](https://docs.pyvista.org/api/core/_autosummary/pyvista.PolyData.delaunay_2d.html) and specify an `edge_source` to properly generate the surface.

**Reply by T4mmi** [reply]
Thanks a lot !
I had issues with `delaunay_2d` besause of unused points (vertices not indexed by faces) 😞 removing them solved it !

**Reply by xiaohulihutu** [reply]
Hi there, thank you for providing such a great library. I have a question regarding to delaunay_2d. If I have a concave polygon that has holes inside of it. When I want to triangulate it, what should it do? I have been trying edge source function but it is not giving what i want. Thank you in advance.

**Reply by whophil** [reply]
@xiaohulihutu I have had issues using VTK's delaunay triangulation with edge sources on polygons with holes. Sometimes it works, sometimes it doesn't. You can find some discussion here: https://pyvista.slack.com/archives/CBA5095UH/p1644372096830669

In the end, I switched to using the `triangle` library (http://www.cs.cmu.edu/~quake/triangle.html), via this Python package (https://github.com/drufat/triangle).

**Reply by xiaohulihutu** [reply]
@whophil Hi sir, thank you very much for your answer, I was testing the triangle library and it works fine the most of time. But the triangulate with "p" arguments may crash on C lib, and I cannot catch any exception in Python. Do you have any idea why? Is it related to the segment or precision of 2d data?

**Reply by whophil** [reply]
@xiaohulihutu I have not encountered that issue, but I would first check for repeated/coincident points in your input data. I do not know if the `triangle` library handles these automatically.
